pulsar-client/src/main/java/org/apache/pulsar/client/impl/ClientCnx.java

```diff
@@ -330,48 +331,54 @@ protected void handleAuthChallenge(CommandAuthChallenge authChallenge) {
         checkArgument(authChallenge.hasChallenge());
         checkArgument(authChallenge.getChallenge().hasAuthData());
 
-        if (Arrays.equals(AuthData.REFRESH_AUTH_DATA_BYTES, authChallenge.getChallenge().getAuthData())) {
-            try {
-                authenticationDataProvider = authentication.getAuthData(remoteHostName);
-            } catch (PulsarClientException e) {
-                log.error("{} Error when refreshing authentication data provider: {}", ctx.channel(), e);
-                connectionFuture.completeExceptionally(e);
-                return;
+        // Run this the code to respond to the authentication challenge in another thread so we are not potentially
+        // blocking a pulsar-client-io thread.  A portion of the code is user supplied, such as getting a token, which is another
+        // reason to run in a separate thread so user code cannot not block pulsar-client-io thread. Potential blocking code
+        // will also render send timeouts to be not respected since it uses pulsar-client-io thread as well.
+        CompletableFuture.runAsync(() -> {
```

Comment:
This will be running from the ForkJoinPool and we don't really know what is running there. We should try to use the internal client executor.

Reply:
+1 to @merlimat suggestion.
we have not control over the internal JVM ForkJoinPool